Add new metadata column to prevent user compaction starvation

cshannon opened the pull request:
This commit adds a new column that will be set by user compactions during the fate operation if it is detected that it can't run due to a system compaction. This will prevent future system compactions from running until the user compaction is finished.

The column is set inside of `CompactionDriver` fate op and then cleaned up by the compaction `CleanUp` fate op. `CompactionCoordinator` has been updated to check the flag on reservation for system compactions.

I'm marking this as a draft for now to get initial feedback (including how things are named) and also because this is untested. There are no tests written for it yet and those need to be added and I haven't tested it out manually either so some stuff could be missing or broken.

This closes #3524

dlmarion commented on server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java:
```diff
@@ -409,7 +409,9 @@ private boolean canReserveCompaction(TabletMetadata tablet, CompactionJob job,
     switch (job.getKind()) {
       case SYSTEM:
-        if (tablet.getSelectedFiles() != null
+        if (tablet.getCompactionRequested()) {
```
If you have not done so already, you will need to trace this back to where the TabletMetadata object is created to ensure that it contains your new column

cshannon replied:
> If you have not done so already, you will need to trace this back to where the TabletMetadata object is created to ensure that it contains your new column

I thought I had done this but I missed that I needed to add the column to be fetched in TabletManagementIterator which was uncovered in my latest test update as it failed until I fixed this.

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/tableOps/compact/CompactionDriver.java:
```diff
@@ -267,8 +267,13 @@ public int updateAndCheckTablets(Manager manager, FateId fateId)
+          // If there are compactions preventing selection of files, then add
           // selecting marker that prevents new compactions from starting
+          var mutator = tabletsMutator.mutateTablet(tablet.getExtent()).requireAbsentOperation()
+              .requireSame(tablet, ECOMP).setCompactionRequested();
```
Its possible that two different fate ops could concurrently set this on tablet.  W/o something to make the key unique one fate op could delete what another set.  One way to avoid this problem is to make the key in the tablet metadata include the fateid, making each fate op have its own key.

```suggestion
              .requireSame(tablet, ECOMP).setCompactionRequested(fateId);
```

cshannon replied:
I initially was going to do this but did not because when I was looking at how the compaction fate operations work I noticed that CompactRange first [reserves](https://github.com/apache/accumulo/blob/da7b3b005e13276c202b4f337496f9ecf2c039e5/server/manager/src/main/java/org/apache/accumulo/manager/tableOps/compact/CompactRange.java#L77) the table with a table lock before continuing to CompactionDriver so I was thinking that would prevent to concurrent fate ops from trying to set it at the same time. How would two different fate ops on the same tablet be able to set this if the table lock is needed to continue and run the CompactionDriver fate op?

keith-turner replied:
> How would two different fate ops on the same tablet be able to set this if the table lock is needed to continue and run the CompactionDriver fate op?

That line mentioned is reserving a "read lock", there is a boolean argument that determines this.  That boolean would be more clear if it were switched to an enum.   The "read lock" allows multiple concurrent compactions to grab the lock as long nothnig holds the write lock (things like create table, delete table, merge grab the write lock).

keith-turner replied:
opened #4276

cshannon replied:
This has been changed in the latest update

keith-turner reviewed (commented):
Not proposing any changes in the following comment.  It was just something I thought about and decided was not worthwhile, but wanted to put up for discussion in case my logic was off.

In addition to the check in CompactionCoordinator we could also add a check to [this section of code](https://github.com/apache/accumulo/blob/10d0824fac42faefba3305d4dcf0dc676dffb54d/server/base/src/main/java/org/apache/accumulo/server/compaction/CompactionJobGenerator.java#L201-L215) to prevent compaction jobs from ever being generated.  The good thing about this check is that avoid polluting the queue with jobs that will not run.  The bad thing about the check is that adds complexity to the code for case that is likely somewhat rare, so based on that it does not seem likes its worth  adding.  I like the existing check in CompactionCoordinator because it filters things at the latest possible time.

cshannon commented:
> Not proposing any changes in the following comment. It was just something I thought about and decided was not worthwhile, but wanted to put up for discussion in case my logic was off.
> 
This is an interesting idea that I hadn't thought of but I agree that for now it is likely not necessary. I would not expect there to be a ton of user compactions spamming things and causing the system jobs to block. However, I am glad you mentioned this and pointed it out because in the future if things change with how the system works or the assumptions are wrong we could go back and add in the check there to prevent creating the jobs in the first place.

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/tableOps/compact/CompactionDriver.java:
```diff
@@ -257,8 +257,13 @@ public int updateAndCheckTablets(Manager manager, FateId fateId)
+              .requireSame(tablet, ECOMP).putUserCompactionRequested(fateId);
+          mutator.submit(tm -> tm.getUserCompactionsRequested().contains(fateId));
+
+          // Add marker here
```
This comment is not related to this code, just this class. At the end of this class there is code to cleanup on failure that could cleanup this new column.

cshannon replied:
I didn't even notice the undo() code, i'll make sure to add that

cshannon replied:
Updated in faf79184b74ac7bbc747f7c8292dc25e7ec01828

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/tableOps/compact/CleanUp.java:
```diff
@@ -74,16 +75,21 @@ public long isReady(FateId fateId, Manager manager) throws Exception {
+          mutator.submit(tabletMetadata -> !tabletMetadata.getCompacted().contains(fateId)
+              && !tabletMetadata.getUserCompactionsRequested().contains(fateId));
           submitted++;
         }
```
Was wondering about the case where user compaction requested column exists but the compacted column does not.  Does not seem like this should happen, was wondering if we should throw an error.  

```suggestion
        } else if(tablet.getUserCompactionsRequested().contains(fateId)){ 
            throw new IllegalStateException("saw user comp req but did not see compacted for <tablet> <fateId>");
       }
```

cshannon replied:
This was refactored to make the checks independent (still one mutation)

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/tableOps/compact/CleanUp.java:
```diff
@@ -74,16 +75,21 @@ public long isReady(FateId fateId, Manager manager) throws Exception {
       t1 = System.nanoTime();
       for (TabletMetadata tablet : tablets) {
         total++;
         if (tablet.getCompacted().contains(fateId)) {
```
This is completely unrelated to this PR, but another comment made me realize that certain compaction metadata is completely unexpected at this point.

```suggestion
        Preconditions.checkState(ensure selected column does not exists)
        if (tablet.getCompacted().contains(fateId)) {
```

cshannon replied:
Should we put this into a follow on PR?

keith-turner commented on test/src/main/java/org/apache/accumulo/test/functional/CompactionIT.java:
```diff
@@ -909,6 +909,82 @@ public void testDeleteCompactionService() throws Exception {
+
+      // Wait and verify that the system compaction finishes
+      Wait.waitFor(() -> {
+        var tabletMeta = ((ClientContext) client).getAmple().readTablet(extent);
```
If files were added before this waitFor then could do some validation in the waitFor

```java
if(!tabletMeta.getUserCompactionsRequested().isEmpty()){
   // if this column exists in the metadata table then it should prevent system compactions from started so do no expect to see any in the metadata table
   assertTrue(tm.getExternalCompactions().values().stream().noneMatch(cm->cm.getKind() == CompactionKind.SYSTEM));
}
```

cshannon replied:
In my latest update I'm looking for CompactionKind.USER instead to verify the user compaction started but not sure if that is as good of a guarantee as this

cshannon commented on test/src/main/java/org/apache/accumulo/test/functional/CompactionIT.java:
```diff
@@ -909,6 +909,82 @@ public void testDeleteCompactionService() throws Exception {
+        var externalCompactions = tabletMeta.getExternalCompactions().size();
+        log.debug("Current external compactions {}", externalCompactions);
+        return externalCompactions == 1;
+      }, Wait.MAX_WAIT_MILLIS, 500);
```
I can do that, the default is actually 1 second, maybe we should make a follow on issue to consider changing the default to something like 100ms as that might be better for most tests

cshannon replied:
Updated in faf79184b74ac7bbc747f7c8292dc25e7ec01828

cshannon commented on test/src/main/java/org/apache/accumulo/test/functional/CompactionIT.java:
```diff
@@ -909,6 +909,82 @@ public void testDeleteCompactionService() throws Exception {
+        return userRequestedCompactions == 0;
+      }, Wait.MAX_WAIT_MILLIS, 500);
+    }
+  }
```
Updated in faf79184b74ac7bbc747f7c8292dc25e7ec01828

cshannon commented:
@keith-turner - This is ready for another review I think, the only outstanding thing is if the test still needs updates. I updated the test to write more data as suggested which should try and trigger another system compaction and I verify that the user compaction runs by checking the CompactionKind but I still am not quite happy with and it and not seeing a good way to verify that the system compaction was indeed blocked as the system compaction thread is a bit non-deterministic in when it runs to look for work so trying to see if there is some state to track but not sure there is. I almost wonder if that might be a better unit test or mocking. Anyways, if you have a chance today you can take a look otherwise I will probably look at the test again tomorrow if I have time.

cshannon commented:
@keith-turner - I think this is good to go now. I updated my test today to check and verify that the next compaction to run after writing more data to trigger a second SYSTEM compaction is a USER compaction and not the SYSTEM compaction that is triggered from the data write (my previous version yesterday only ensured the USER compaction eventually ran but did not guarantee it was the next one). 

Well, when I did this my test started to fail and was showing another SYSTEM compaction kept running. I started debugging and noticed that every time the check was done inside `CompactionCoordinator` to see if there were pending requested user compactions it was false even after it was set to true by the fate operation. I couldn't figure out why and finally found it in `TabletManagementIterator`...I forgot to add the new marker column to the list of columns to fetch, once I did that it started to read it and things now work as expected where the SYSTEM compaction gets blocked by the pending USER compaction when the test is run and if you change the fate operation to no longer set the marker then the test fails because Accumulo runs another SYSTEM compaction instead of the USER compaction.

cshannon commented:
@keith-turner - After I fixed the issue with the column `USER_COMPACTION_REQUESTED` not being read by the scanner inside `TabletManagementIterator` I realized that the `ensureFetched()` check didn't throw an exception like I would have expected and I didn't know why because I looked and there are already unit tests I added that show that check does work and throws an exception if missing.

So I investigated and it is because the column was added to the fetchedCols inside TabletManagement by adding it to the `CONFIGURED_COLUMNS` set [here](https://github.com/apache/accumulo/pull/4254/files#diff-a5e21c00133e7d52c9de49a1a42583ac55f7257daab36825638ed4dbbf0d9acbR48). That meant that when the [getUserCompactionsRequested()](https://github.com/apache/accumulo/pull/4254/files#diff-f7cd33f7e1f9321ed285551c95577effc1f35880f80dcaa6e1c74d1ccae8987eR412) was called it didn't throw an illegal state exception since the column was added to the fetched set. 

However, we it still didn't work until the actual column was [added](https://github.com/apache/accumulo/pull/4254/files#diff-1adf517589f978e7f9cae95a7fdb404a52db469edf102500c3e2885ea7c0eb77R141) to the scanner in my latest commit. I'm thinking there must be a better way to handle that or I missed something because the state check is saying that it was fetched but it wasn't with how it's set up so wondering if there is a better way to do this so it would fail if the scanner is missing the column.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/functional/CompactionIT.java:
```diff
@@ -909,6 +912,124 @@ public void testDeleteCompactionService() throws Exception {
+
+        if (current.isPresent()) {
+          var currentCompaction = current.orElseThrow();
+          log.debug("Current running compaction {}", currentCompaction.getKey());
```
I think this log stmt is redundant.

keith-turner commented:
>  I'm thinking there must be a better way to handle that or I missed something because the state check is saying that it was fetched but it wasn't with how it's set up so wondering if there is a better way to do this so it would fail if the scanner is missing the column.

Kinda related, I had made [this comment](https://github.com/apache/accumulo/blob/58be770cb54331081965ef5689e5722a0d47797e/server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java#L127) and thought I had opened an issue but could not find it.  The code has evolved a lot from main and is fetching a lot more columns. I think we could remove all of the fetches and get everything for the tablet.   That would simplify things for the problem you ran into.

cshannon commented on server/base/src/main/java/org/apache/accumulo/server/metadata/ConditionalTabletMutatorImpl.java:
```diff
@@ -231,6 +232,13 @@ private void requireSameSingle(TabletMetadata tabletMetadata, ColumnType type) {
         mutation.addCondition(c);
       }
         break;
+      case USER_COMPACTION_REQUESTED: {
```
Good catch on the missing test for this and also for MetadatConstraints. I was focused mostly on the CompactionIT and forgot about looking for other tests we have to update

cshannon commented:
> Kinda related, I had made [this comment](https://github.com/apache/accumulo/blob/58be770cb54331081965ef5689e5722a0d47797e/server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java#L127) and thought I had opened an issue but could not find it. The code has evolved a lot from main and is fetching a lot more columns. I think we could remove all of the fetches and get everything for the tablet. That would simplify things for the problem you ran into.

That makes sense, I'm thinking that might be best to change as a separate follow on to track/test it separately.

keith-turner commented:
> That makes sense, I'm thinking that might be best to change as a separate follow on to track/test it separately.

Opened #4307
